#975 Part 2: Context params used as source for factory and lifecycle methods

agudian opened the pull request:
Only the last commit is new here, the first commits are identical to what's open in PR #1006.

I've already worked in the changes in the Javadoc of `Context` and `ObjectFactory`. I'll do the adaptation of `BeforeMapping` / `AfterMapping` seperately in a PR for #992, if that's okay.

The introduction of `@Context` in the ref-docs will be included in this PR later on (probably after christmas), but I wanted to show you how it's coming along. Code-wise, I'm finished with it so you could play around if you like.

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java:
```diff
@@ -111,7 +111,6 @@ else if ( Executables.isGetterMethod( accessor ) ) {
         protected TargetWriteAccessorType targetWriteAccessorType;
         protected Type targetType;
         protected Accessor targetReadAccessor;
-        protected TargetWriteAccessorType targetReadAccessorType;
```
Was marked as unused, so I removed it.

agudian commented:
Commit is rebased to the current master and ready for review.

One thing you might want to think about: as it is implemented now, any non-void methods without parameters on the context parameter type are considered as potential factory methods (as we do it on mappers and `uses` types as well). So it's consistent in that way, but perhaps users want to have context types with such methods (e.g. `getFirstCreatedInstance()`) or want to use context param types that are not under their control that by chance have such methods. That's why I had to change `FactoryContext#createNode` in the test to not collect it as potential factory method...

So should we make `@ObjectFactory` mandatory for factory methods on `@Context` parameter types?

agudian commented:
> Commit is rebased to the current master and ready for review.

uhm, what I mean with "ready for review" is that the code parts are done, but the ref-doc update is still missing. I'll start on that soonish and add it to this PR. 😁

agudian commented:
I've added the Javadoc changes to `@BeforeMapping` and `@AfterMapping`, and added the reference-documentation.

filiphr commented on core-common/src/main/java/org/mapstruct/Context.java:
```diff
@@ -25,14 +25,22 @@
- * For generated code to call a method that is declared with {@link Context} parameters, the declaration of the mapping
- * method being generated needs to contain at least those (or assignable) {@link Context} parameters as well. MapStruct
- * will not create new instances of missing {@link Context} parameters nor will it pass {@code null} instead.
+ * The type of an {@code @Context} parameter is also inspected for factory methods (with or without the {@code @}
```
Nothing special, but why not use `{@code @}{@link Context}` like the other annotations?

agudian replied:
I didn't want to link if we are already within that file... 🤓

filiphr replied:
oh yea you are right. I didn't see that I was reading in that file 😄

filiphr commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
+
+*Note:* no `null` checks are performed before calling factory or before/after mapping methods on context parameters. The caller needs to make sure that no `null` are passed in that case.
+
+For generated code to call a method that is declared with `@Context` parameters, the declaration of the mapping method being generated needs to contain at least those (or assignable) `@Context` parameters as well. MapStruct will not create new instances of missing`@Context` parameters nor will it pass `null` instead.
```
a space is missing in 
> missing`@Context`

filiphr commented:
Regarding your question @agudian, I think that maybe the others can help more then me.

And I forgot to add a note to me. The changes in the `IterableMappingMethod` need to be taken into consideration in #964.

gunnarmorling commented:
WDYM by

> without parameters on the context parameter type

?

gunnarmorling commented:
> context param types that are not under their control that by chance have such methods

In which case are methods declared by a context parameter considered at all? So far I understood the context parameter(s) would just be passed through, but you seem to imply that their methods would potentially be invoked by generated code?

agudian commented:
> non-void methods without parameters on the context parameter type

That means a method without parameters that returns something other than `void` is declared on the type of the context parameter. We usually identify such methods as candidate factory methods (in mappers and types added with `@Mapper(uses = ...)`.

agudian commented:
> So far I understood the context parameter(s) would just be passed through, but you seem to imply that their methods would potentially be invoked by generated code?

That's the whole point of this PR 👍. You can have `@BeforeMapping` / `@AfterMapping` and `@ObjectFactory` methods declared on the type of the context parameter. Check the type `SelfContainingCycleContext` in the tests and it's usages for one example of what to do with that.

agudian commented:
Oh, I also added that as "Example 2" to the Javadoc of `@interface Context` :)

filiphr commented:
@agudian, when you will be doing the rebase over the master. Can you please make sure that the `StreamMappingMethod` has the new things applied in the `IterableMappingMethod` as well.

Once I do #1012 then it will be easier and changes that actually affect both methods will be in once place

gunnarmorling commented on core-common/src/main/java/org/mapstruct/Context.java:
```diff
@@ -78,6 +80,61 @@
+ *
+ *     NodeDto nodeDto = new NodeDto();
+ *
+ *     cycleContext.storeMappedInstance( node, nodeDto );
```
How is it ensured that `getMappedInstance()` is called on line 113 and here then `storeMappedInstance()`?

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
+
+Additional _context_ or _state_ information can be passed through generated mapping methods to custom methods with `@Context` parameters. Such parameters are passed to other mapping methods, `@ObjectFactory` methods (see <<object-factories>>) or `@BeforeMapping` / `@AfterMapping` methods (see <<customizing-mappings-with-before-and-after>>) when applicable and can thus be used in custom code.
+
+The type of an `@Context` parameter is also inspected for factory methods and `@BeforeMapping` / `@AfterMapping` methods, which are called on the provided context parameter value if applicable.
```
"`@Context` parameters are also searched for..."

But I'm not quite convinced of the overall idea.

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
+
+*Note:* no `null` checks are performed before calling factory or before/after mapping methods on context parameters. The caller needs to make sure that no `null` are passed in that case.
+
+For generated code to call a method that is declared with `@Context` parameters, the declaration of the mapping method being generated needs to contain at least those (or assignable) `@Context` parameters as well. MapStruct will not create new instances of missing `@Context` parameters nor will it pass `null` instead.
```
Can you put one sentence per line? That makes it easier to do diffs down the road.

"MapStruct will" sounds as if it s a runtime component. Better change it into something like "Code generated by MapStruct will not...".

agudian replied:
I'd gladly do so, I just followed the old habit from the awestruct days. Would it be okay if I did that for the whole file at once, in a seperate commit or PR? Then my text editor can do it for me and we have it fixed once and for all... 😁

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -1490,8 +1538,7 @@ By default, the generated code for mapping one bean type into another will call
 
-To make use of custom factories register them via `@Mapper#uses()` as described in <<invoking-other-mappers>>. When creating the target object of a bean mapping, MapStruct will look for a parameterless method, 
-a method annotated with `@ObjectFactory`, or a method with only one `@TargetType` parameter that returns the required target type and invoke this method instead of calling the default constructor:
+To make use of custom factories register them via `@Mapper#uses()` as described in <<invoking-other-mappers>> or add a `@Context` parameter (see <<passing-context>>). When creating the target object of a bean mapping, MapStruct will look for a parameterless method, a method annotated with `@ObjectFactory`, or a method with only one `@TargetType` parameter that returns the required target type and ...
```
Better make it one sentence per line.

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/context/SelfContainingFactoryContext.java:
```diff
@@ -0,0 +1,49 @@
+import org.mapstruct.ap.test.context.NodeDto.AttributeDto;
+
+/**
+ * A type to be used as {@link Context} parameter to create NodeDTO and AttributeDTO instances
```
DTO=Dto.

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/context/ContextParameterTest.java:
```diff
@@ -77,15 +81,35 @@ public void automappingWithContextCorrectlyResolvesCycles() {
+    @Test
+    public void automappingWithSelfContainingContextCorrectlyResolvesCycles() {
+        Node root = buildNodes();
+        NodeDto rootDto = AutomappingNodeMapperWithSelfContainingContext.INSTANCE
```
What does "Automapping" mean?

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/context/AutomappingNodeMapperWithSelfContainingContext.java:
```diff
@@ -0,0 +1,40 @@
+    AutomappingNodeMapperWithSelfContainingContext INSTANCE =
+        Mappers.getMapper( AutomappingNodeMapperWithSelfContainingContext.class );
+
+    NodeDto nodeToNodeDto(Node node, @Context SelfContainingCycleContext cycleContext,
```
Should we really support multiple context parameters? I can see issues or at least very complex rules around method selection. A single context parameter should be enough, users can structure them as needed.

gunnarmorling replied:
@agudian Seems you never got back to this one?

gunnarmorling commented:
> I admit the cycle detection use case is compelling, but I find the feature very tough to grok and use. Rules around it seem quite involved.

Specifically, could that use case not be solved by simply invoking the right methods on the context object from within `@BeforeMapping` methods?

agudian commented:
> Specifically, could that use case not be solved by simply invoking the right methods on the context object from within @BeforeMapping methods?

Yes it can be, and one of the existing test cases already demonstrates that. But I find this for some use cases to be a much better solution, as it is totally self-contained (data and logic in the same place) and I imagine that to be an easier solution to hand out for users wanting to handle cycles (and perhaps a starting point for automatically handling it in our code)

gunnarmorling commented:
> Would it be okay if I did that for the whole file at once, in a seperate
commit or PR?

+1 for a separate PR. Maybe combining it with splitting the guide into one
file per chapter?

2017-01-10 22:20 GMT+01:00 Andreas Gudian <notifications@github.com>:

> *@agudian* commented on this pull request.
> ------------------------------
>
> In documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc
> <https://github.com/mapstruct/mapstruct/pull/1016>:
>
> > @@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
>  ----
>  ====
>
> +[[passing-context]]
> +=== Passing context or state objects to custom methods
> +
> +Additional _context_ or _state_ information can be passed through generated mapping methods to custom methods with `@Context` parameters. Such parameters are passed to other mapping methods, `@ObjectFactory` methods (see <<object-factories>>) or `@BeforeMapping` / `@AfterMapping` methods (see <<customizing-mappings-with-before-and-after>>) when applicable and can thus be used in custom code.
> +
> +The type of an `@Context` parameter is also inspected for factory methods and `@BeforeMapping` / `@AfterMapping` methods, which are called on the provided context parameter value if applicable.
> +
> +*Note:* no `null` checks are performed before calling factory or before/after mapping methods on context parameters. The caller needs to make sure that no `null` are passed in that case.
> +
> +For generated code to call a method that is declared with `@Context` parameters, the declaration of the mapping method being generated needs to contain at least those (or assignable) `@Context` parameters as well. MapStruct will not create new instances of missing `@Context` parameters nor will it pass `null` instead.
>
> I'd gladly do so, I just followed the old habit from the awestruct days.
> Would it be okay if I did that for the whole file at once, in a seperate
> commit or PR? Then my text editor can do it for me and we have it fixed
> once and for all... 😁
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mapstruct/mapstruct/pull/1016>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABvxMhOWUi_G_z9byMObXPbT1untRmZks5rQ_YygaJpZM4LVB5O>
> .
>

sjaakd commented:
I start to like the idea. A while ago I created a stageful mapper PR. The idea there was that we could implement handwritten methods (e.g. in an abstract-class-mapper) and use the context in these mappings. Use cases were:

1. I have 2 modes (quality regimes) in which a mapper should work.  The archive regime has a bigger set of possible values in general for a property than the new regime (more restricted). So take property X it maps String value A to K, B to L, C to M under the archive regime. But under the standard regime A maps to K, B to N, C to N. All other 'value' properties are mapped in a similar way. I would normally require a hand-written method to do each value mapping (with qualifiers pointing out which property it concerns (each one is string-to-string). In Andreas his solution I would do this with an `@AfterMapping` and hand the complete object. The context would have the complete object and do all stuff since it knows the quality regime. All value properties would be ignored.

2. ID generation. Some properties require a unique ID (counter) in a web service request. I now handle this with a ThreadLocal and an expression. I would use a IdGenerator (wrapping a counter) as context. Now I would again use a post processing method and call the getId (incrementing counter and return Id) on the IdGenerator. All concerned Id properties would be ignored.

These scenarios are mixed. So I would require 2 context objects, since their purpose is entirely different (nice separation of concerns). That covers (I believe one of your other remarks @gunnarmorling above).


So to make a long story short: I have already 2 use cases that I can cover quite easily with this solution. :smile:

sjaakd commented:
See my case 1 above: @agudian .. Is it possible to have an interface as `@Context` object and stick the annotation `@AfterMapping` to the interface?

agudian commented:
Absolutely! 🙂

sjaakd commented:
Nice 👍

agudian commented:
I've rebased and improved the documentation bits as Gunnar suggested, but I'll leave the line-breaks and split-ups for #991.
@gunnarmorling, I guess you'll need the weekend to think a bit more about it? Would a beer help? 🍻  😉

gunnarmorling commented:
@agudian thinking a bit more about this one, would it suffice to have support for lifecycle methods (before-mapping, after-mapping) on context objects? But factory methods and other mapping methods would be obtained from "used" mappers as before?

agudian commented:
Sure, it's a start. 😄 
Leaving out the factory methods seems a bit arbitrary to me, though. So I'd like to understand the rationale.

Is it because of the dilemma with the non-annotated factory methods I described above? I'd sleep better without using non-annotated factory methods as well.

But methods explicitly annotated with `@ObjectFactory`? Wouldn't it be strange to support the `@BeforeMapping`/`@AfterMapping` methods, but not `@ObjectFactory`? Or is it more strange if we would require `@ObjectFactory` on context parameters for factory methods without arguments, but leave them optional on _used_ mappers?

Or is it because we can't devise a proper life-like example? 😁

gunnarmorling commented:
Hey @agudian, sorry for the delay, finally coming back to this.

So my preference for only supporting lifecycle methods (before-mapping, after-mapping) on context objects is due to the fact that i'd expect those methods to be the ones actually interacting with the context. I.e. I don't see a strong reason for opening up more ways where mapping (and factory) methods should be defined; in a library, there should be one way to achieve a given goal. Your question around `@ObjectFactory` is a good example, why creating this issue if we could simply stick to the current ways of defining mapping and factory methods?

I guess one could say the same about the lifecycle methods, but here I find the use case more compelling indeed (as it allows for "plugging in" a context with its logic). So my personal preference is be to limit it to this - unless we can find a really good use case for the alternative.

We still can expand it in subsequent revisions if people ask for it.

agudian commented:
Okay, PR is updated to not search for factory methods on `@Context` params anymore. I did some rewriting of the `@ObjectFactory` javadoc that I left in... I didn't mention factory methods *on* context params there anyway.

gunnarmorling commented on core-common/src/main/java/org/mapstruct/Context.java:
```diff
@@ -25,14 +25,22 @@
+ * will not create new instances of missing {@code @Context} parameters nor will it pass {@code null} instead.
+ * <p>
+ * <strong>Example 1:</strong> Using {@code @Context} parameters for passing data down to hand-written property mapping
+ * methods and {@code @} {@link BeforeMapping} methods:
```
Minor detail: the "@" seems to be handled a bit inconsistently. So far we always expressed it via "&#64;". Here I probably wouldn't bother adding it at all.

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
+
+`@Context` parameters are also searched for `@BeforeMapping` / `@AfterMapping` methods, which are called on the provided context parameter value if applicable.
+
+*Note:* no `null` checks are performed before calling before/after mapping methods on context parameters. The caller needs to make sure that no `null` are passed in that case.
```
...to make sure that `null` is not passed...

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
+[[passing-context]]
+=== Passing context or state objects to custom methods
+
+Additional _context_ or _state_ information can be passed through generated mapping methods to custom methods with `@Context` parameters. Such parameters are passed to other mapping methods, `@ObjectFactory` methods (see <<object-factories>>) or `@BeforeMapping` / `@AfterMapping` methods (see <<customizing-mappings-with-before-and-after>>) when applicable and can thus be used in custom code.
```
Can we stick to one sentence per line? Makes it easier with diffs down the road.

agudian replied:
I'd like to reformat that whole file with #991... It's even assigned to me already... 😁

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -924,6 +924,54 @@ public class CarMapperImpl implements CarMapper {
+[source, java, linenums]
+[subs="verbatim,attributes"]
+----
+public abstract CarDto toCar(Car car, @Context Locale translationLocale);
```
Nice example!

agudian commented:
> Seems you never got back to this one?

What do you mean? (sorry, can't comment up there, I don't get a reply box for the last comment)

gunnarmorling commented:
My question was about whether multiple context parameters should be supported or just a single one. I can see more complexities coming in when doing the former. E.g. say your original method takes tour context parameters of the same type and calls another one which only takes one parameter of that same type. Which one is passed on?

agudian commented:
> My only remaining question is whether we should limit it to one context parameter. May make reasoning about ordering and so on a bit simpler.

I could imagine some use cases where you would want to handle different aspects in different branches of the data structure... I think Sjaak even had an example for this? As long as the types are different, there's no real problem here anyway. But I'd like to hear if users come back with use cases where they have more than one context parameter of the same type (e.g. different `Locale` parameters for different parts of their beans) and then see what we could do there - if it comes up.

agudian commented:
> E.g. say your original method takes tour context parameters of the same type and calls another one which only takes one parameter of that same type. Which one is passed on?

Ah, okay - currently you get an error if you declare a method with more than one context param of the same type.

agudian commented:
See `ContextParameterErroneousTest` (was part of the Part 1 PR)...
